src/core/src/pattern/op/optional.cpp

```diff
@@ -43,6 +43,7 @@ bool ov::pass::pattern::op::Optional::match_value(Matcher* matcher,
     // Turn the Optional node into WrapType node to create a case where the Optional node is present
     ov::OutputVector input_values_to_optional = input_values();
     size_t num_input_values_to_optional = input_values_to_optional.size();
+    bool same_type = pattern_value.get_element_type() == graph_value.get_element_type();
```

Comment:
Did not get the usage of `type` in non-related type pattern.. So, could you please explain?

Reply:
Imagine you're matching 2 nodes like that. Please, pay attention that the pattern has no inputs.
![image](https://github.com/openvinotoolkit/openvino/assets/109364720/ee3b03c8-310b-4208-9fe0-78789e4405ff)

With the current implementation of `if (matcher->match_value(pattern, graph_value) || num_input_values_to_optional == 0) {`
even if `matcher->match_value()` fails (as it should in this case), we will still hit the case of `num_input_values_to_optional == 0` resulting in the nodes matching which is wrong. That's why I suggest checking for a number of inputs to optional being zero **_only_** if  the types of nodes match.

Reply:
Anyway, a bit strange solution in mix with previous PR. I will check it myself in my branch.

Reply:
We have removed it as it's not really working. But the case it a valid scenario in which the current implementation is failing.